File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/Jump.java
package com.only4.algorithm.version4j.leetcode;

public class Jump {
    public int jump(int[] nums) {
        if (nums.length <= 1) return 0;

        int currentReach = 0;
        int maxReach = 0;
        int jumps = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            maxReach = Math.max(maxReach, i + nums[i]);

            if (i == currentReach) {
                currentReach = maxReach;
                jumps++;

                if (currentReach >= nums.length - 1) {
                    return jumps;
                }
            }
        }
        return jumps;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/jump.kt
package com.only4.algorithm.version4k.leetcode

fun jump(nums: IntArray): Int {
    if (nums.size <= 1) return 0

    var currentReach = 0
    var maxReach = 0
    var jumps = 0

    for (i in 0 until nums.size - 1) {
        maxReach = maxOf(maxReach, i + nums[i])

        if (i == currentReach) {
            currentReach = maxReach
            jumps++

            if (currentReach >= nums.lastIndex) {
                return jumps
            }
        }
    }
    return jumps
}
